### PR TITLE
Make timezone table independent from ORC

### PR DESCRIPTION
**Pull request by vuule**
## Description
Modifies the creation of timezone transition tables to make it usable outside of the ORC context:

- Remove the ORC epoch offset from the transition table.
- Use `timestamp_s` and `duration_s` instead of integral values in the table.
- Return table as `cudf::table` instead of the special `timezone_table` struct.
- Move the implementation/header out of `cudf::io::orc`.
- Split the header into C++ and CUDA parts. C++ API has public and `detail` counterparts.

Other:
Adapt ORC reader to changes to the timezone transition table.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by ttnghia** [comment]
I see that the timezone stuff is in the `cudf/orc/` folder? I believe that they are very independent, so please consider refactoring it out into somewhere else not in `cudf/orc/`, like `cudf/utilities/`. This would be useful for implementing other timezone-aware operations.

**Comment by vuule** [comment]
> I see that the timezone stuff is in the `cudf/orc/` folder? I believe that it is very independent from ORC file format, so please consider refactoring it out into somewhere else not in `cudf/orc/`, like `cudf/utilities/`. This would be useful for implementing other timezone-aware operations.

This PR was a first draft just to unblock @shwina on the Python side. I've since split the headers to separate C++ and CUDA code and moved them to /include/detail. The namespace was also changed to cudf::detail.

**Comment by vuule** [comment]
CC @shwina to confirm that new API and header location are suitable for Cython use.

**Comment by shwina** [comment]
@vuule Thanks! while technically the new location is OK for Cython, can the function be moved out of the `detail` namespace?

**Comment by vuule** [comment]
> @vuule Thanks! while technically the new location is OK for Cython, can the function be moved out of the `detail` namespace?

I placed it under `detail` because it's not really intended for end users. This is based on an offline discussion with @galipremsagar . I'm fine with moving it out of detail, the `detail` namespace is only meant to discourage direct use here.

**Comment by shwina** [comment]
I think any code that is used outside of `libcudf` should be in its public API (even if that is just cuDF Python).

This avoids situations in the future where someone might decide to drop this code from `libcudf` because it's no longer being used internally, OR, a breaking API change is made but not considered breaking because it's not part of the public API. 

Perhaps it could be excluded from the doxygen-generated docs, so that libcudf consumers do not discover it?

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

File: cpp/include/cudf/timezone.hpp
Comment (shwina) [inline cpp/include/cudf/timezone.hpp:47]: Should this argument have a default value of `stream_default`?

Comment (shwina) [reply cpp/include/cudf/timezone.hpp:47]: Actually, I think this argument should just be eliminated from the public API

Comment (mythrocks) [reply cpp/include/cudf/timezone.hpp:47]: I've lost track of where we stand on eventually exposing `stream` parameters from non-`detail` public functions. Would removing the parameter mean having to add it back eventually?
Perhaps adding a default value is an acceptable compromise?

File: cpp/include/cudf/detail/timezone.cuh
Comment (mythrocks) [inline cpp/include/cudf/detail/timezone.cuh:63]: ELI5: What's the calculation here? I'm afraid I don't follow.

File: cpp/src/io/orc/stripe_enc.cu
Comment (mythrocks) [inline cpp/src/io/orc/stripe_enc.cu:52]: This constant looks similar to the `duration_s` in `stripe_data.cu`. I'm not sure I've understood why we have two constants for the same value.

File: cpp/include/cudf/timezone.hpp
Comment (shwina) [reply cpp/include/cudf/timezone.hpp:47]: cc: @vyasr

File: cpp/include/cudf/detail/timezone.cuh
Comment (vuule) [reply cpp/include/cudf/detail/timezone.cuh:63]: It codifies the leap year rules:


In our Gregorian calendar three criteria must be taken into account to identify leap years:
The year must be evenly divisible by 4;
If the year can also be evenly divided by 100, it is not a leap year;
unless...
The year is also evenly divisible by 400. Then it is a leap year.

I believe this translates to 100 (every fourth year) - 4 (every hundredth year ) + 1 (every four hundredth year) leap days in 400 years.

File: cpp/include/cudf/timezone.hpp
Comment (vuule) [reply cpp/include/cudf/timezone.hpp:47]: yeah, now that I look at this as a regular libcudf API, it should not have a stream parameter. I'll add a `detail` counterpart that has the defaulted stream parameter.

File: cpp/src/io/orc/stripe_enc.cu
Comment (vuule) [reply cpp/src/io/orc/stripe_enc.cu:52]: it is the same constant, just for the reader and the writer. I'll try to move it to a common header.

File: cpp/include/cudf/timezone.hpp
Comment (vyasr) [reply cpp/include/cudf/timezone.hpp:47]: Yeah I would remove the stream. As for where we stand, the answer is eventually but with no concrete timeline yet. We're making incremental progress towards it but for now I think the only public APIs exposing streams are ones that a privileged user (typically cuspatial) has explicitly requested as necessary.

File: cpp/include/cudf/detail/timezone.cuh
Comment (bdice) [inline cpp/include/cudf/detail/timezone.cuh:43]: Can we give this a longer name? Does `ttimes` mean `transition_times`? Let's try to increase clarity as much as possible in names.

File: cpp/include/cudf/detail/timezone.cuh
Comment (bdice) [reply cpp/include/cudf/detail/timezone.cuh:63]: Is this a rule we can get (or reference) from some other source? If we must implement it, is this codified in a way that is specific to ORC conventions or general to all time zone libraries?

To the greatest extent possible, cuDF should avoid implementing or defining its own rules for time math. We should always use (or reference) `libcudacxx`, other timezone libraries, an RFC, or similar.

File: cpp/include/cudf/timezone.hpp
Comment (bdice) [inline cpp/include/cudf/timezone.hpp:40]: Let's not use an obsolete time zone for our example. `US/Pacific` has been replaced by `America/Los_Angeles`.

http://www.timezoneconverter.com/cgi-bin/zoneinfo.tzc?s=default&tz=US/Pacific

**Review by bdice — CHANGES_REQUESTED** [review]
Suggestions on time zone rules and naming. (Haven't dug much into the actual C++)

File: cpp/include/cudf/detail/timezone.cuh
Comment (vuule) [reply cpp/include/cudf/detail/timezone.cuh:63]: This is the definition of the Gregorian calendar. It's unrelated to ORC or timezones.
Given how specific this is (only multiples of 400 years have a fixed number of days, otherwise you need to know the year range) I doubt any library can compute this without computing number of days for each year individually (1-400 in this case, I think).

I can add a link to Gregorian calendar and clarification comments.

Comment (bdice) [reply cpp/include/cudf/detail/timezone.cuh:63]: I think I'm asking something different. I want a reference that states why we "search the 400-year cycle if outside of the file entries range" (in the comment below). We didn't come up with that, it's defined by some spec outside of libcudf.

Comment (vuule) [reply cpp/include/cudf/detail/timezone.cuh:63]: Ah, that. I actually don't know where this comes from. I'll try to find it and incorporate in some way.

**Comment by vuule** [comment]
Looks like I managed to straight-up murder this branch. Will open a new PR when I find the changes.

File: cpp/include/cudf/detail/timezone.cuh
Comment (vuule) [reply cpp/include/cudf/detail/timezone.cuh]: So, I've been digging around the interwebs and could not find anything on the use of 400 year cycle to compute time zone offsets in the future. 
The only thing I found is the name for this cycle - [Solar cycle](https://en.wikipedia.org/wiki/Solar_cycle_(calendar)). And even that is a seemingly overloaded term.
I added this to the names for the relevant constants, will incorporate it here as well.

File: cpp/include/cudf/detail/timezone.cuh
Comment (vuule) [reply cpp/include/cudf/detail/timezone.cuh]: Made the name more specific, but kept the `snake_case` style. I hope it's at least a bit easier to see that it's a constant, based on the name.

File: cpp/include/cudf/detail/timezone.cuh
Comment (vuule) [reply cpp/include/cudf/detail/timezone.cuh:29]: My understanding is that UT, UTC and GMT are equivalent for `second` precision. I have changed the TZif related code to use UT, but since ORC references UTC, I kept it there as such.

File: cpp/include/cudf/detail/timezone.cuh
Comment (vuule) [reply cpp/include/cudf/detail/timezone.cuh]: tried to make this clearer. Please let me know if it's still lacking

**Review by shwina — APPROVED** [review]
LGTM from Cython side!

File: cpp/include/cudf/timezone.hpp
Comment (bdice) [inline cpp/include/cudf/timezone.hpp:30]: How do we know there are two entries per year? Is it guaranteed by the spec that timezones cannot have zero, one, three, or four entries?

**Review by bdice — COMMENTED** [review]
Code looks good. One minor question about the far, far future.

File: cpp/include/cudf/timezone.hpp
Comment (vuule) [reply cpp/include/cudf/timezone.hpp:30]: There are two entries because of Daylight Saving Time. If the time zone does not have DST, all future entries will have the same time offset. We could optimize this case in the future.

Comment (bdice) [reply cpp/include/cudf/timezone.hpp:30]: It seems that we’re hardcoding an assumption that daylight savings time exists and results in exactly two changes per year. I don’t think either of those are safe assumptions.

Comment (vuule) [reply cpp/include/cudf/timezone.hpp:30]: It is a safe assumption, since that's how `make_timezone_transition_table` works (currently). It will always include two entires per year in cycle. The case when DST does not exist, it will create 2 X cycle_years of the same offset, which is still correct, just slower than necessary.
We could separate the output columns into tzif entires and future entries. With that, we can have any number of elements in the future column. I would like to keep optimizations like this to a future PR. Let me know what you think.

Comment (bdice) [reply cpp/include/cudf/timezone.hpp:30]: Document this ("it always includes two entries per year, assuming two transitions per year due to Daylight Savings Time") and any potential future work in a comment. That would be enough to resolve this and move forward.

Comment (vuule) [reply cpp/include/cudf/timezone.hpp:30]: Agree with the suggestion, this on not clear from the API.
Added more detailed description of the transition table.

File: cpp/include/cudf/timezone.hpp
Comment (bdice) [inline cpp/include/cudf/timezone.hpp:38]:
```suggestion
 * last transition, the table includes entries that form a `solar_cycle_years`-year cycle (future
```

**Review by bdice — APPROVED** [review]
Typo fix. Otherwise LGTM!

**Pull request merged**
